core/server/src/main/java/alluxio/worker/netty/FileDataServerHandler.java

```diff
@@ -12,19 +12,115 @@
-   * @param conf the configuration to use
+   * @param configuration the configuration to use
+   */
+  public FileDataServerHandler(FileSystemWorker worker, Configuration configuration) {
+    mWorker = worker;
+  }
+
+  /**
+   * Handles a {@link RPCFileReadRequest} by reading the data through an input stream provided by
+   * the file worker. This method assumes the length to read is less than or equal to the unread
+   * data in the file.
+   *
+   * @param ctx The context of this request which handles the result of this operation
+   * @param req The initiating {@link RPCFileReadRequest}
+   * @throws IOException if an I/O error occurs when interacting with the UFS
+   */
+  public void handleFileReadRequest(ChannelHandlerContext ctx, RPCFileReadRequest req)
+      throws IOException {
+    req.validate();
+
+    long ufsFileId = req.getTempUfsFileId();
+    long offset = req.getOffset();
+    long length = req.getLength();
+    byte[] data = new byte[(int) length];
+
+    // TODO(calvin): This can be more efficient for sequential reads if we keep state
+    try (InputStream in = mWorker.getUfsInputStream(ufsFileId, offset)) {
+      int read = in.read(data);
+      if (read != length) {
+        LOG.error("Only read " + read + " bytes of data when " + length + " was requested.");
+        throw new IOException(); // Intend to catch below
+      }
+      RPCFileReadResponse resp =
+          new RPCFileReadResponse(ufsFileId, offset, length, new DataByteBuffer(
+              ByteBuffer.wrap(data), length), RPCResponse.Status.SUCCESS);
+      ChannelFuture future = ctx.writeAndFlush(resp);
+      future.addListener(ChannelFutureListener.CLOSE);
+    } catch (Exception e) {
+      LOG.error("Failed to read ufs file", e);
+      RPCFileReadResponse resp =
+          RPCFileReadResponse.createErrorResponse(req, RPCResponse.Status.UFS_READ_FAILED);
+      ChannelFuture future = ctx.writeAndFlush(resp);
+      future.addListener(ChannelFutureListener.CLOSE);
+    }
+  }
+
+  /**
+   * Handles a {@link RPCFileWriteRequest} by writing the data through an output stream provided
+   * by the file worker. This method only allows appending data to the file and does not support
+   * writing at arbitrary offsets.
+   *
+   * @param ctx The context of this request which handles the result of this operation
+   * @param req The initiating {@link RPCFileWriteRequest}
+   * @throws IOException if an I/O error occurs when interacting with the UFS
    */
-  public FileDataServerHandler(FileSystemWorker worker, Configuration conf) {
+  public void handleFileWriteRequest(ChannelHandlerContext ctx, RPCFileWriteRequest req)
+      throws IOException {
+    long ufsFileId = req.getTempUfsFileId();
+    // Currently unused as only sequential write is supported
+    long offset = req.getOffset();
+    long length = req.getLength();
+    final DataBuffer data = req.getPayloadDataBuffer();
 
+    try {
+      OutputStream out = mWorker.getUfsOutputStream(ufsFileId);
```

Comment:
The thrift call `FileSystemWorker.completeUfsFile` or `FileSystemWorker.cancelUfsFile`